warp10/src/main/java/io/warp10/continuum/gts/GTSHelper.java

```diff
@@ -2190,117 +2249,412 @@ public static final GeoTimeSerie bucketize(GeoTimeSerie gts, long bucketspan, in
+    Map<String, String> labels = gts.getLabels();
 
     //
     // Loop on all buckets
     //
 
-    //
-    // We can't skip buckets which are before the first tick or after the last one
-    // because the bucketizer function might set a default value when it encounters an
-    // empty sub serie
-    //
+    if (memoryOptimized) {
+      // to keep compatibility, BUCKETIZE will still browse buckets from the most recent to the oldest
+      // a bucket is all values between (bucketEnd - bucketspan + 1) and bucketEnd.
 
-    // Allocate a stable GTS instance which we will reuse when calling subserie
-    GeoTimeSerie subgts = null;
-
-    for (int i = 0; i < bucketcount; i++) {
-
-      long bucketend = lastbucket - i * bucketspan;
-
-      //
-      // Extract GTS containing the values that fall in the bucket
-      // Keep multiple values that fall on the same timestamp, the
-      // aggregator functions will deal with them.
-      //
+      // early exit if input is empty, or bucketize time window is obviously not in input
+      if (gts.values == 0 || firsttick > lastbucket || lasttick < (lastbucket - bucketspan * bucketcount + 1)) {
+        // either input is empty, either the last bucket is before the first tick.
+        return bucketized;
+      }
 
-      subgts = subSerie(gts, bucketend - bucketspan + 1, bucketend, false, false, subgts);
+      // sort input
+      GTSHelper.sort(gts);
 
-      if (0 == subgts.values) {
-        continue;
+      // find array index of last bucket, if needed
+      int i;
+      if (lastbucket > lasttick) {
+        i = gts.size() - 1;
+      } else {
+        i = Arrays.binarySearch(gts.ticks, 0, gts.values, lastbucket);
+        if (-1 == i) {
+          // should not be there, this case leads to early exit before.
+          return bucketized;
+        } else if (i < 0) {
+          // just before the insertion point
+          i = -i - 1 - 1;
+        }
       }
 
-      Object[] aggregated = null;
-
+      // first case: bucketizer is a macro. We need to build a subgts for each bucket and expose it on the stack.
+      // building a subgts will lead to new allocation and array copy. TODO: find an alternative to new allocations here
       if (null != stack) {
         if (!(aggregator instanceof Macro)) {
           throw new WarpScriptException("Expected a macro as bucketizer.");
         }
 
+        // size and gts internal arrays will be overriden anyway
+        GeoTimeSerie subgts = new GeoTimeSerie(1);
         subgts.safeSetMetadata(bucketized.getMetadata());
-        stack.push(subgts);
-        stack.exec((Macro) aggregator);
+        subgts.type = gts.type;
+        subgts.sorted = true; // we will copy sorted data from gts.
+        subgts.reversed = false;
 
-        Object res = stack.peek();
+        // iterate on input to find buckets
+        long currentBucketEnd;
+        int currentBucketEndPosition;
+        int currentBucketStartPosition;
```

Comment:
the name aren't extremely clear at first read, can you add some comments ?
int currentBucketEndPosition; // index of the newest tick in the bucket
int currentBucketStartPosition; // index of the oldest tick in the bucket